README.md
Welcome to _Namecoin Wallet_, a standalone Namecoin payment app for your Android device!

This project contains several sub-projects:

 * __wallet__:
     The Android app itself. This is probably what you're searching for.
 * __market__:
     App description and promo material for the Google Play app store.
 * __integration-android__:
     A tiny library for integrating digitial payments into your own Android app
     (e.g. donations, in-app purchases).
 * __sample-integration-android__:
     A minimal example app to demonstrate integration of digital payments into
     your Android app.

You can build all sub-projects at once using Maven:

`mvn clean install`


wallet/CHANGES
Namecoin Wallet
4.14
* Added Checkpoints
*

v0.1-beta
* forked for Namecoin
-----------------------
Groestlcoin Wallet
v0.22-beta
* fixed paper wallet sweep

v0.21-beta
* changed to beta
* updated icons

v0.20-alpha
* updated to Bitcoin Wallet 4.14

digitalcoin Wallet
-----------------
v3.01.3561
* Fixed Version 3 fork bug
* updated checkpoints

v3.01.3550
* updated for Digitalcoin Core 3.0.1
* Raised Transaction Fee's to 0.1 DGC to support miners

v0.2.1-beta
* updated checkpoints
* turned off bloom filters

v0.2.0-beta
* updated checkpoints
* updated to Bitcoin Wallet 3.56 code (sweep, fixes, translations)

v0.1.7-beta
* updated checkpoints
* added DNS nodes from Digitalcoin Core 2.0 source

v0.1.6-beta
* Bug fix associated with Peer Discovery

v0.1.5-beta
* update to bitcoin-wallet 3.46
* added Litecoin's PeerDBDiscovery System
* added IRC bootstrap for more peers
* Send Coins shows the transaction fee (currently the minimum fee)

v0.1.4-beta
* permanent fix for bugs relating to pause activities (removed temp fixes and support-v4r13)

v0.1.3-beta
* fixed several bugs relating to pausing activities

v0.1.2-beta
* fixed checkpoints to work for the Version 1.0 blocks (between 476280 and 523800)
* added two nodes from a list of addnodes from the bitcointalk topic

v0.1.1-beta
* renamed to hashengineering.digitalcoin.wallet
* added support for importing Bitcoin Private Keys from first digitalcoin app (Crypto-Expert)

v1.0.1.5
* Fixed menu item doubling on rotate in address screen - langerhans (https://github.com/langerhans/dogecoin-wallet-new)
* replaced block explorer with blockr.io instead of the apparently dead dgc.cryptocoinexplorer.com
* Added BTER as a backup for getting exchange rates
* added mDGC as an exchange rate.
* added Disconnect feature back in
* updated to bitcoin-wallet v3.39

v1.0.1.4
* Updated to v3.31 (bitcoin-wallet)
* Few bug fixes (import parser)

v1.0.1.3

* Added checkpoints file for faster blockchain sync
* Sync with bitcoin-wallet 3.30

v1.0.1.2

* Fix broken diff adjust in 1.0.1.1

v1.0.1.1

*  updated digitalcoin 1.0.1 - difficulty adjustment fork

v1.00
* Forked bitcoin-wallet v3.23 for digitalcoin V1.00 (http://digitalcoin.co)

Bitcoin Wallet
=======
v4.14

* Update Turkish, Norwegian Bokmål, Japanese, French and Danish translations.
* Small bugfixes.

v4.13

* Support sweeping of password-protected paper wallets (BIP38).
* Update Swedish translation.
* Update android-maven-plugin to 4.0.0, so the build should now be compatible to newer Maven versions.
* Update to ProGuard 5.1, making it possible to build with a Java 8 compiler.

v4.12

* Small bugfixes.

v4.11

* When sending coins, there is a "quick confirm" option which pays the old higher fee.
* Fix white/black screen that prevents some users from starting the app.
* Fix reuse of change address in some cases.
* Based on bitcoinj 0.12.2.

v4.10

* Fix sent transaction not updating on send coins screen.
* Fix compatibility issue to GMail 5 app when sending report attachments.
* Migrate clipboard to use Android 4 API.

v4.09

* Migrate preferences and NFC to use Android 4 API.

v4.08

* Update Norwegian Bokmål, Japanese, Danish, Finnish, Spanish, Catalan and Russian translations.

v4.07

* Performance improvements.
* Drop network fee by factor of 10.
* Support rotation for HD key chains.
* Based on bitcoinj 0.12.1.

v4.06

* The wallet can be protected by a spending PIN.
* Record memos read from payment requests.

v4.05

* Re-add textual Bitcoin address to enlarged QR code.

v4.04

* Small bugfixes.

v4.03

* Remove textual Bitcoin address from main screen.
* Warning for users with pay-to-many transactions. Those will slow down their
  wallet.
* Workaround for crash in the network monitor.

v4.02

* Fix crash when sweeping wallets.
* New translation into Catalan.
* Update Korean, Russian, Italian and Danish translations.

v4.01

* New translation into Norwegian Bokmål.
* Small bugfix.

v4.0

* All Bitcoin addresses are now deterministically derived from one single secret,
  which was created when you installed the app for the first time. So from now on,
  only one backup is needed to recover all your future funds in that wallet.
* The app does its best to not reuse addresses any more. "Request coins" will use a
  fresh address on each invocation. The current address in the QR code on the main
  page changes as soon as coins are received on it.
* When sending coins, the fee is shown while you type and before you press "Send".
  After sending, you'll see the paid fee in the transaction list.
* When sending coins, most errors are now shown already before you press Send.
* The app now requires Android 4.0.3 or higher. This makes UI development much
  easier in future.
* The app has a fresher, sharper looking app icon. The icon for the testnet version
  is now entirely grey.
* On Android 4.1 or higher, the app-widget is now horizontally resizable.
* Based on bitcoinj 0.12.

v3.59

* Resolve issue with restoring backups that have been renamed.
* Update Turkish, Korean, Ukrainian, Russian, Italian, French, Finnish and Danish translations.
* Remove BROADCAST_STICKY permission.

v3.58

* Small bugfixes.

v3.57

* App-widget: show approximate balance in local currency.

v3.55-v3.56

* New translation into Ukrainian. Update many other translations.
* Small bugfixes.

v3.54

* Support compressed keys when sweeping a wallet.
* Update Russian translation.

v3.53

* New translation into Finnish.
* Update Czech, Korean, Turkish, Spanish, Japanese, Hebrew and Italian translations.

v3.52

* Update French, Danish and Korean translations.

v3.51

* Sweeping of paper wallets (e.g. those used for cold storage).
* New translation into Korean.
* Small bugfixes.

v3.50

* Update Chinese, Danish, Italian and Japanese translations.
* Update QR-code scanner to ZXing 3.1.0.
* Small bugfixes.
* Based on bitcoinj 0.11.3.

v3.49

* Update French and Turkish translations.
* Small bugfixes.

v3.48


wallet/src/de/schildbach/wallet/Constants.java
/*
 * Copyright 2011-2014 the original author or authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.schildbach.wallet;

import android.os.Build;
import android.os.Environment;
import android.text.format.DateUtils;
import hashengineering.namecoin.wallet.R;
import org.bitcoinj.core.CoinDefinition;
import org.bitcoinj.core.NetworkParameters;
import org.bitcoinj.params.MainNetParams;
import org.bitcoinj.params.TestNet3Params;
import org.bitcoinj.utils.MonetaryFormat;

import java.io.File;

/**
 * @author Andreas Schildbach
 */
public final class Constants
{
	public static final boolean TEST = R.class.getPackage().getName().contains("_test");

	/** Network this wallet is on (e.g. testnet or mainnet). */
	public static final NetworkParameters NETWORK_PARAMETERS = TEST ? TestNet3Params.get() : MainNetParams.get();

	public final static class Files
	{
		private static final String FILENAME_NETWORK_SUFFIX = NETWORK_PARAMETERS.getId().equals(NetworkParameters.ID_MAINNET) ? "" : "-testnet";

		/** Filename of the wallet. */
		public static final String WALLET_FILENAME_PROTOBUF = "wallet-protobuf" + FILENAME_NETWORK_SUFFIX;

		/** Filename of the automatic key backup (old format, can only be read). */
		public static final String WALLET_KEY_BACKUP_BASE58 = "key-backup-base58" + FILENAME_NETWORK_SUFFIX;

		/** Filename of the automatic wallet backup. */
		public static final String WALLET_KEY_BACKUP_PROTOBUF = "key-backup-protobuf" + FILENAME_NETWORK_SUFFIX;

		/** Manual backups go here. */
		public static final File EXTERNAL_WALLET_BACKUP_DIR = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS);

		/** Filename of the manual key backup (old format, can only be read). */
		public static final String EXTERNAL_WALLET_KEY_BACKUP = CoinDefinition.coinName.toLowerCase() +"bitcoin-wallet-keys" + FILENAME_NETWORK_SUFFIX;

		/** Filename of the manual wallet backup. */
		public static final String EXTERNAL_WALLET_BACKUP = CoinDefinition.coinName.toLowerCase() +"-wallet-backup" + FILENAME_NETWORK_SUFFIX;

		/** Filename of the block store for storing the chain. */
		public static final String BLOCKCHAIN_FILENAME = "blockchain" + FILENAME_NETWORK_SUFFIX;

		/** Filename of the block checkpoints file. */
		public static final String CHECKPOINTS_FILENAME = "checkpoints" + FILENAME_NETWORK_SUFFIX + ".txt";
	}

	/** Maximum size of backups. Files larger will be rejected. */
	public static final long BACKUP_MAX_CHARS = 10000000;

	private static final String EXPLORE_BASE_URL_PROD = CoinDefinition.BLOCKEXPLORER_BASE_URL_PROD;
	private static final String EXPLORE_BASE_URL_TEST = CoinDefinition.BLOCKEXPLORER_BASE_URL_TEST;
	/** Base URL for browsing transactions, blocks or addresses. */
	public static final String EXPLORE_BASE_URL = NETWORK_PARAMETERS.getId().equals(NetworkParameters.ID_MAINNET) ? EXPLORE_BASE_URL_PROD
			: EXPLORE_BASE_URL_TEST;
    public static final String EXPLORE_ADDRESS_PATH  = CoinDefinition.BLOCKEXPLORER_ADDRESS_PATH;
    public static final String EXPLORE_TRANSACTION_PATH  = CoinDefinition.BLOCKEXPLORER_TRANSACTION_PATH;
    public static final String EXPLORE_BLOCK_PATH  = CoinDefinition.BLOCKEXPLORER_BLOCK_PATH;


	//public static final String MIMETYPE_PAYMENTREQUEST = "application/"+ CoinDefinition.coinTicker.toLowerCase() +"-paymentrequest"; // BIP 71
	//public static final String MIMETYPE_PAYMENT = "application/"+ CoinDefinition.coinTicker.toLowerCase() +"-payment"; // BIP 71
	//public static final String MIMETYPE_PAYMENTACK = "application/"+ CoinDefinition.coinTicker.toLowerCase() +"-paymentack"; // BIP 71

	public static final String MIMETYPE_BACKUP_PRIVATE_KEYS = "x-"+CoinDefinition.coinName.toLowerCase()+"/private-keys";

	private static final String BITEASY_API_URL_PROD = CoinDefinition.UNSPENT_API_URL;//"https://api.biteasy.com/blockchain/v1/";
	private static final String BITEASY_API_URL_TEST = "https://api.biteasy.com/testnet/v1/";
	/** Base URL for blockchain API. */
	public static final String BITEASY_API_URL = NETWORK_PARAMETERS.getId().equals(NetworkParameters.ID_MAINNET) ? BITEASY_API_URL_PROD
			: BITEASY_API_URL_TEST;

	/** URL to fetch version alerts from. */
	public static final String VERSION_URL = "http://wallet.schildbach.de/version";

	/** MIME type used for transmitting single transactions. */
	public static final String MIMETYPE_TRANSACTION = "application/x-"+CoinDefinition.coinTicker.toLowerCase()+"tx";

	/** MIME type used for transmitting wallet backups. */
	public static final String MIMETYPE_WALLET_BACKUP = "application/x-bitcoin-wallet-backup";

	/** Number of confirmations until a transaction is fully confirmed. */
	public static final int MAX_NUM_CONFIRMATIONS = 7;

    /** User-agent to use for network access. */
	public static final String USER_AGENT = CoinDefinition.coinName +" Wallet";

	/** Default currency to use if all default mechanisms fail. */

	public static final String DEFAULT_EXCHANGE_CURRENCY = "USD";

    /** Currency code for base 1 Bitcoin. */
	public static final String CURRENCY_CODE_BTC = CoinDefinition.coinTicker;

    /** Currency code for base 1/1000 Bitcoin. */
	public static final String CURRENCY_CODE_MBTC = "m" + CoinDefinition.coinTicker;

    /** Currency code for base 1/1000000 Bitcoin. */
	public static final String CURRENCY_CODE_UBTC = "µ" + CoinDefinition.coinTicker;

	public static final String DONATION_ADDRESS = CoinDefinition.DONATION_ADDRESS;;

	/** Recipient e-mail address for reports. */
	public static final String REPORT_EMAIL = "hashengineeringsolutions@gmail.com";;

	/** Subject line for manually reported issues. */
	public static final String REPORT_SUBJECT_ISSUE = "Reported issue";

	/** Subject line for crash reports. */
	public static final String REPORT_SUBJECT_CRASH = "Crash report";

	public static final char CHAR_HAIR_SPACE = '\u200a';
	public static final char CHAR_THIN_SPACE = '\u2009';
	public static final char CHAR_ALMOST_EQUAL_TO = '\u2248';
	public static final char CHAR_CHECKMARK = '\u2713';
	public static final char CURRENCY_PLUS_SIGN = '\uff0b';
	public static final char CURRENCY_MINUS_SIGN = '\uff0d';
	public static final String PREFIX_ALMOST_EQUAL_TO = Character.toString(CHAR_ALMOST_EQUAL_TO) + CHAR_THIN_SPACE;
	public static final int ADDRESS_FORMAT_GROUP_SIZE = 4;
	public static final int ADDRESS_FORMAT_LINE_SIZE = 12;


	public static final String LICENSE_URL = "http://www.gnu.org/licenses/gpl-3.0.txt";

    public static final String FORKED_FROM_SOURCE = "based on bitcoin-wallet 4.14\n";
    public static final String FORKED_FROM_SOURCE_BITCOINJ = "based on bitcoinj 0.12\n";
	public static final String CREDITS_BITCOINJ_URL = "https://github.com/HashEngineering/" + CoinDefinition.coinName.toLowerCase() + "j";
	public static final String CREDITS_ZXING_URL = "http://code.google.com/p/zxing/";
    public static final String CREDITS_WEBSITE_URL = "http://namecoin.org/";
    public static final String CREDITS_FORUM_URL = "http://namecoin.org/forum/";

	public static final String CREDITS_ICON_URL = "https://bitcointalk.org/index.php?action=profile;u=2062";

	public static final String AUTHOR_TWITTER_URL = "https://twitter.com/#!/HashEngineering";
	public static final String AUTHOR_GOOGLEPLUS_URL = "https://profiles.google.com/HashEngineering";

	public static final String COMMUNITY_GOOGLEPLUS_URL = "https://plus.google.com/communities/105515929887248493912";


	public static final String MARKET_PUBLISHER_URL = "market://search?q=pub:\"Hash Engineering Solutions\"";
//=======
	public static final MonetaryFormat LOCAL_FORMAT = new MonetaryFormat().noCode().minDecimals(3).optionalDecimals();

    public static final String SOURCE_URL = "https://github.com/HashEngineering/" + CoinDefinition.coinName.toLowerCase() + "-wallet";
    public static final String BINARY_URL = "https://github.com/HashEngineering/"+ CoinDefinition.coinName.toLowerCase() +"-wallet/releases";

    public static final String MARKET_APP_URL = "market://details?id=%s";
    public static final String WEBMARKET_APP_URL = "https://play.google.com/store/apps/details?id=%s";//>>>>>>> origin/master

	public static final int HTTP_TIMEOUT_MS = 15 * (int) DateUtils.SECOND_IN_MILLIS;
	public static final int PEER_TIMEOUT_MS = 15 * (int) DateUtils.SECOND_IN_MILLIS;

	public static final long LAST_USAGE_THRESHOLD_JUST_MS = DateUtils.HOUR_IN_MILLIS;
	public static final long LAST_USAGE_THRESHOLD_RECENTLY_MS = 2 * DateUtils.DAY_IN_MILLIS;

	public static final int SDK_JELLY_BEAN = 16;
	public static final int SDK_JELLY_BEAN_MR2 = 18;

	public static final int SDK_DEPRECATED_BELOW = Build.VERSION_CODES.ICE_CREAM_SANDWICH;

	public static final boolean BUG_OPENSSL_HEARTBLEED = Build.VERSION.SDK_INT == Constants.SDK_JELLY_BEAN
			&& Build.VERSION.RELEASE.startsWith("4.1.1");

	public static final int MEMORY_CLASS_LOWEND = 48;
}
